[stdlib] Fix String.reserveCapacity underallocation

natecook1000 opened the pull request:
When called on a string that is not uniquely referenced, `String.reserveCapacity(_:)` ignores the current capacity, using the passed-in capacity for the size of its new storage. This can result in an underallocation and write past the end of the new buffer.

This fix changes the new size calculation to use the current UTF-8 count as the minimum. ~At the same time, it removes a growth factor from that calculation, as the appropriate growth is already being factored in upstream at all call sites.~

rdar://109275875
Fixes #53483

natecook1000 committed:
[stdlib] Fix String.reserveCapacity underallocation

When called on a string that is not uniquely referenced,
`String.reserveCapacity(_:)` ignores the current capacity, using
the passed-in capacity for the size of its new storage. This can
result in an underallocation and write past the end of the new
buffer.

This fix changes the new size calculation to use the current UTF-8
count as the minimum. At the same time, it removes a growth factor
from that calculation, as the appropriate growth is already being
factored in upstream at all call sites.

rdar://109275875
Fixes #53483

natecook1000 commented:
@swift-ci Please test

natecook1000 commented:
@swift-ci Please benchmark

natecook1000 commented:
@swift-ci Please test

natecook1000 commented:
@swift-ci Please benchmark

natecook1000 commented:
@swift-ci Please benchmark

kperryua commented on stdlib/public/core/StringGutsRangeReplaceable.swift:
```diff
@@ -94,20 +94,28 @@ extension _StringGuts {
-    let growthTarget = Swift.max(n, (self.uniqueNativeCapacity ?? 0) * 2)
+    // We're non-unique OR smaller than requested, so as we copy to new storage,
+    // choose a size that at least stores the current contents.
+    let growthTarget = Swift.max(n, self.utf8Count)
```
I was trying to decide if in the case of unique native storage whether should we also lower-bound this by the current capacity. But since we’re definitely creating a copy at this point, that doesn’t seem necessary. So this LGTM.

natecook1000 replied:
That was my initial thought, but we actually [have a test codifying the current behavior](https://github.com/apple/swift/blob/main/test/stdlib/NewStringAppending.swift#L136-L148), which shrinks the new storage to fit the size of the contents. Didn't want to end up with unexpectedly large capacities, so now this better matches what we're doing now.

lorentey commented on stdlib/public/core/StringGutsRangeReplaceable.swift:
```diff
@@ -94,20 +94,28 @@ extension _StringGuts {
-    let growthTarget = Swift.max(n, (self.uniqueNativeCapacity ?? 0) * 2)
+    // We're non-unique OR smaller than requested, so as we copy to new storage,
+    // choose a size that at least stores the current contents.
+    let growthTarget = Swift.max(n, self.utf8Count)
```
This is not right -- the  `* 2` currently serves a crucial role, and removing it without reviewing and updating all callers would result in an algorithmic slowdown.

`grow` is currently being called from both `reserveCapacity`/`init(_initialCapacity:)` _and_ `replaceSubrange`/`append`. In the `reserveCapacity` case, we want to allocate precisely as much memory as desired (still accepting if malloc gives us more, of course), while in the `replaceSubrange` case we **must** grow storage exponentially (if the requested capacity is still within the same order of magnitude than the current size).

For example, the loop below is heavily relying on this logic so that it will only reallocate storage O(log(_n_)) times:

```swift
var s = ""
for _ in 0 ..< n {
  s.insert("a", at: s.endIndex)
}
```

Getting rid of the doubling would make this loop resize string storage O(_n_) times, rendering its runtime performance accidentally quadratic instead of linear.

(I'm using `insert` in the example because `append` is a special case: some of its code paths have their own exponential resizing code, _in addition to_ calling `grow`. `String`'s storage sizing logic has become way too obfuscated over the years; it could use a full audit and cleanup.)

lorentey reviewed (changes requested):
This isn't quite right -- we need to be careful to preserve the expected performance of loops that e.g. build strings by appending smaller pieces.

lorentey commented on validation-test/stdlib/String.swift:
```diff
@@ -2370,4 +2370,23 @@ StringTests.test("SmallString.zeroTrailingBytes") {
+    var copy = str
+    copy.reserveCapacity(20)
+    copy.append(contentsOf: String(repeating: "z", count: 50))
+  }
```
It would be nice to document why these two tests exist, and/or to add some actual checks to them. (For example, we could check that `copy` has the expected contents at the end.)

Why does this test has a loop?

natecook1000 committed:
Maintain existing string growth strategy

This change maintains the existing doubling of capacity when
growing a unique native buffer. Non-native or non-unique strings
only allocate the requested capacity (or space enough for the
current contents, if that's larger than what's requested).

natecook1000 commented:
@swift-ci Please test

natecook1000 commented:
@lorentey Restored the growth factor inside `_StringGuts.grow` and improved the tests — thank you!

natecook1000 commented:
@swift-ci Please benchmark

natecook1000 commented:
I don't see anything alarming in the benchmarks:

```
------- Performance (x86_64): -O -------

IMPROVEMENT                           OLD       NEW       DELTA    RATIO    
FlattenListLoop                       1623.0    1450.0    -10.7%   **1.12x (?)**
Breadcrumbs.MutatedUTF16ToIdx.Mixed   211.636   191.667   -9.4%    **1.10x (?)**
Breadcrumbs.MutatedIdxToUTF16.Mixed   218.875   198.778   -9.2%    **1.10x (?)**
CxxStringConversion.cxx.to.swift      159.667   148.25    -7.2%    **1.08x (?)**

------- Performance (x86_64): -Osize -------

REGRESSION                               OLD       NEW        DELTA    RATIO    
ArrayAppendGenericStructs                1244.0    1753.333   +40.9%   **0.71x (?)**
FlattenListLoop                          1426.0    1631.0     +14.4%   **0.87x (?)**
String.replaceSubrange.ArrChar.Small     36.828    39.87      +8.3%    **0.92x (?)**
String.replaceSubrange.Substring.Small   39.769    43.04      +8.2%    **0.92x (?)**

IMPROVEMENT                              OLD       NEW        DELTA    RATIO    
Breadcrumbs.MutatedUTF16ToIdx.Mixed      212.1     191.3      -9.8%    **1.11x (?)**
Breadcrumbs.MutatedIdxToUTF16.Mixed      218.875   198.444    -9.3%    **1.10x (?)**

------- Performance (x86_64): -Onone -------

REGRESSION                               OLD       NEW       DELTA    RATIO    
NSStringConversion.InlineBuffer.ASCII    5261.0    6114.0    +16.2%   **0.86x (?)**
CharacterLiteralsLarge                   383.4     440.0     +14.8%   **0.87x (?)**
NSStringConversion.InlineBuffer.UTF8     3186.0    3616.0    +13.5%   **0.88x (?)**
RemoveWhereFilterInts                    1801.0    1953.0    +8.4%    **0.92x (?)**
String.replaceSubrange.Substring.Small   41.231    44.5      +7.9%    **0.93x (?)**

IMPROVEMENT                              OLD       NEW       DELTA    RATIO    
ArrayAppendGenericStructs                1410.0    1240.0    -12.1%   **1.14x (?)**
Breadcrumbs.MutatedUTF16ToIdx.Mixed      223.333   202.182   -9.5%    **1.10x (?)**
Breadcrumbs.MutatedIdxToUTF16.Mixed      230.429   210.0     -8.9%    **1.10x (?)**
```
